### PR TITLE
Metadata schema provider from cluster configuration

### LINKED ISSUE #67: Add a new MetadataSchemaProvider that receives the connection data through a conf file
We should allow the creation of a `MetadataSchemaProvider` that receives a properties files with the cluster configuration.

### PR DESCRIPTION
**Pull request by fedefernandez**
Fixes #67 

This PR adds a new builder method for the MetadataSchemaProvider based on a Cluster Configuration.

@raulraja, could you take a look? I'm using some `free` types and not sure if this is the right way to do it.

### CODE REVIEW

File: core/src/main/scala/schema/provider/MetadataSchemaProvider.scala [hunk 1]
```diff
@@ -52,14 +58,15 @@ class MetadataSchemaProvider[M[_]](cluster: Cluster)(implicit H: FSHandler[Clust
         M.flatMap(finalizer)(_ => e.fold(M.raiseError, M.pure))
```
Comment (juanpedromoreno) [inline core/src/main/scala/schema/provider/MetadataSchemaProvider.scala]: I know that's unrelated but in this scope, we have `M` as MonadError evidence, which matches with the class type constructor `M[_]`. Maybe we could rename it, wdyt?

File: core/src/main/scala/schema/provider/MetadataSchemaProvider.scala [hunk 2]
```diff
@@ -52,14 +58,15 @@ class MetadataSchemaProvider[M[_]](cluster: Cluster)(implicit H: FSHandler[Clust
...
 
-    def closeF[F[_]](implicit clusterAPI: ClusterAPI[F]): FreeS[F, Unit] =
-      clusterAPI.close
+    M.flatMap(clusterProvider) { cluster =>
+      implicit val H: FSHandler[ClusterAPI.Op, M] =
+        clusterAPIHandler[M] andThen apiInterpreter[M, Cluster](cluster)
 
-    M.flatMap(guarantee(metadataF[ClusterAPI.Op].interpret[M], closeF[ClusterAPI.Op].interpret[M])) {
-      metadata =>
+      M.flatMap(guarantee(metadataF.interpret[M], closeF.interpret[M])) { metadata =>
```
Comment (juanpedromoreno) [inline core/src/main/scala/schema/provider/MetadataSchemaProvider.scala]: Here we have `metadataF` and `closeF` with two different interpretations. Can we compose both and interpret just once?

File: core/src/test/scala/schema/provider/MetadataSchemaProviderSpec.scala [hunk 3]
```diff
@@ -110,9 +101,41 @@ class MetadataSchemaProviderSpec
...
+          MetadataSchemaProvider.clusterProvider[Future](reader),
+          scala.concurrent.duration.Duration.Inf)
+
+        Option(cluster.getClusterName) shouldBe validClusterConfiguration.name
+      }
+
     }
 
+    "create a metadataSchemaProvider from a Reader with the configuration" in {
+      val reader: Reader = new StringReader("cluster = {}")
```
Comment (juanpedromoreno) [inline core/src/test/scala/schema/provider/MetadataSchemaProviderSpec.scala:132]: Can we add more test cases for different meta-schemas? makes sense?

**Review by juanpedromoreno — APPROVED** [review]
I've added minor comments/questions, other than that, LGTM.

File: core/src/main/scala/schema/provider/MetadataSchemaProvider.scala (hunk 2 above)
Comment (fedefernandez) [reply core/src/main/scala/schema/provider/MetadataSchemaProvider.scala]: They are two different operations for the `guarantee` method, but I'll take a look

File: core/src/test/scala/schema/provider/MetadataSchemaProviderSpec.scala (hunk 3 above)
Comment (fedefernandez) [reply core/src/test/scala/schema/provider/MetadataSchemaProviderSpec.scala:132]: I didn't take a look to the coverage, let me iterate over it to see what can I add

Comment (fedefernandez) [reply core/src/test/scala/schema/provider/MetadataSchemaProviderSpec.scala:132]: The problem with different configurations is that the validation is made by the builder in the Datastax driver so we'll not be testing our code. Also, I think it's well covered right now. The only missing parts are untestable code due to final, private Java classes restrictions.

[See MetadataSchemaProvider.scala coverage](https://codecov.io/gh/frees-io/freestyle-cassandra/pull/75/src/core/src/main/scala/schema/provider/MetadataSchemaProvider.scala?before=core/src/main/scala/schema/provider/MetadataSchemaProvider.scala)

**Review by juanpedromoreno — APPROVED** [review]

**Pull request merged**